tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SavedModelBundle.java

```diff
@@ -89,11 +110,107 @@ private Loader(String exportDir) {
+     * has a single main graph and one or more signatures). These models are compatible with those
+     * exported by TensorFlow 1.x or by TensorFlow 2.x estimators.
+     *
+     * <p>Therefore, all functions exported in a model should share the same session at the moment
+     * or an exception will be thrown.</i>
+     *
+     * @param function a function carrying a signature and a valid session to the graph to be saved
+     * @return this object
+     * @throws IllegalArgumentException if a function with the same name has already been added to the model
+     */
+    public Exporter withFunction(ConcreteFunction function) {
+      Signature signature = function.signature();
+      if (functions.containsKey(signature.key())) {
+        throw new IllegalArgumentException("Function \"" + signature.key() + "\" was already added to the model");
+      }
+      functions.put(signature.key(), function);
+      if (session == null) {
+        session = function.session();
+      } else if (session != function.session()) {
+        throw new UnsupportedOperationException("Saving multiple functions with different graphs/sessions is not supported yet.");
+      }
+      metaGraphDefBuilder.putSignatureDef(signature.key(), signature.asSignatureDef());
+      return this;
+    }
+
+    /**
+     * Save the model into the export directory.
+     *
+     * @throws IOException if saved model or variable state can be written on disk
+     */
+    public void export() throws IOException {
+      if (functions.isEmpty() || session == null) {
+        throw new IllegalStateException("Model should contain at least one valid function");
+      }
+      if (tags.isEmpty()) {
+        tags.add(DEFAULT_TAG);
+      }
+      // It is imperative to retrieve the graphDef after the saverDef, as the former might add
+      // new ops to the graph for saving and restoring the variables.
+      Graph graph = session.graph();
+      MetaGraphDef.Builder metaGraphDef = metaGraphDefBuilder
+          .setSaverDef(graph.saverDef())
+          .setGraphDef(graph.toGraphDef())
+          .setMetaInfoDef(MetaInfoDef.newBuilder().addAllTags(tags));
+      functions.forEach((k, f) -> metaGraphDef.putSignatureDef(k, f.signature().asSignatureDef()));
+
+      // Make sure saved model directories exist
+      Path variableDir = Paths.get(exportDir, "variables");
+      variableDir.toFile().mkdirs();
+
+      // Save the variables state
+      session.save(variableDir.resolve("variables").toString());
+
+      // Save the graph
+      SavedModel savedModelDef = SavedModel.newBuilder().addMetaGraphs(metaGraphDef).build();
+      try (OutputStream file =
+          new FileOutputStream(Paths.get(exportDir, "saved_model.pb").toString())) {
+        savedModelDef.writeTo(file);
+      }
+    }
+
+    Exporter(String exportDir) {
+      this.exportDir = exportDir;
+    }
+
+    private final String exportDir;
+    private final List<String> tags = new ArrayList<>();
```

Comment:
Calling `tags()` on the exporter is cumulative, while calling it on the loader always resets the previous value. I agree that we can make this consistent, anyway tags in general are not very useful since most models only have a single tag, as suggested by TensorFlow.